src/Native/Unix/System.Native/pal_networking.c

```diff
@@ -1915,10 +1943,17 @@ static bool TryConvertSocketTypePalToPlatform(int32_t palSocketType, int* platfo
     }
 }
 
-static bool TryConvertProtocolTypePalToPlatform(int32_t palProtocolType, int* platformProtocolType)
+static bool TryConvertProtocolTypePalToPlatform(int32_t palAddressFamily, int32_t palProtocolType, int* platformProtocolType)
 {
     assert(platformProtocolType != NULL);
 
+    if (palAddressFamily == AddressFamily_AF_NETLINK || palAddressFamily == AddressFamily_AF_PACKET || palAddressFamily == AddressFamily_AF_CAN)
+    {
+        // Protocol type is specific to AddressFamily. For non-IP protocols just pass what ever caller give us and let OS to handle it.
```

Comment:
In all existing cases the .NET supplied value is platform-agnostic, right? But in these new cases, the value is platform-specific? What happens if, for example, macOS starts supporting these options? How would we maintain compatibility while also supporting am OS that may use different values?

Reply:
I think we would need to than map Linux values to other OSes in this function. I was thinking about adding enum for AddressFamily specific protocols but that can be somewhat large and hard to keep complete.

Reply:
Are any of the existing ProtocolTypes valid with any of these new AddressFamilies?  It just seems risky to say that the ProtocolType is a PAL value except when X, Y, or Z is the AddressFamily, in which case it's a raw platform-specific value.

Reply:
yes, the protocols do overlap - at least at OS level. For Netlink, one can also register own sub-protocol. It is rare so we may not care. 

```
in.h:
enum {
  IPPROTO_IP = 0,       /* Dummy protocol for TCP       */
#define IPPROTO_IP      IPPROTO_IP
  IPPROTO_ICMP = 1,     /* Internet Control Message Protocol    */
#define IPPROTO_ICMP        IPPROTO_ICMP
  IPPROTO_IGMP = 2,     /* Internet Group Management Protocol   */
#define IPPROTO_IGMP        IPPROTO_IGMP
  IPPROTO_IPIP = 4,     /* IPIP tunnels (older KA9Q tunnels use 94) */
#define IPPROTO_IPIP        IPPROTO_IPIP
  IPPROTO_TCP = 6,      /* Transmission Control Protocol    */
#define IPPROTO_TCP     IPPROTO_TCP
  IPPROTO_EGP = 8,      /* Exterior Gateway Protocol        */
#define IPPROTO_EGP     IPPROTO_EGP
  IPPROTO_PUP = 12,     /* PUP protocol             */
#define IPPROTO_PUP     IPPROTO_PUP
  IPPROTO_UDP = 17,     /* User Datagram Protocol       */
#define IPPROTO_UDP     IPPROTO_UDP
  IPPROTO_IDP = 22,     /* XNS IDP protocol         */
#define IPPROTO_IDP     IPPROTO_IDP
  IPPROTO_TP = 29,      /* SO Transport Protocol Class 4    */
#define IPPROTO_TP      IPPROTO_TP
  IPPROTO_DCCP = 33,        /* Datagram Congestion Control Protocol */
...

in6.h

#define IPPROTO_HOPOPTS     0   /* IPv6 hop-by-hop options  */
#define IPPROTO_ROUTING     43  /* IPv6 routing header      */
#define IPPROTO_FRAGMENT    44  /* IPv6 fragmentation header    */
#define IPPROTO_ICMPV6      58  /* ICMPv6           */
#define IPPROTO_NONE        59  /* IPv6 no next header      */
#define IPPROTO_DSTOPTS     60  /* IPv6 destination options */
#define IPPROTO_MH      135 /* IPv6 mobility header     */

netlink.h

#define NETLINK_ROUTE       0   /* Routing/device hook              */
#define NETLINK_UNUSED      1   /* Unused number                */
#define NETLINK_USERSOCK    2   /* Reserved for user mode socket protocols  */
#define NETLINK_FIREWALL    3   /* Unused number, formerly ip_queue     */
#define NETLINK_SOCK_DIAG   4   /* socket monitoring                */
#define NETLINK_NFLOG       5   /* netfilter/iptables ULOG */
#define NETLINK_XFRM        6   /* ipsec */
#define NETLINK_SELINUX     7   /* SELinux event notifications */
#define NETLINK_ISCSI       8   /* Open-iSCSI */
#define NETLINK_AUDIT       9   /* auditing */
#define NETLINK_FIB_LOOKUP  10
#define NETLINK_CONNECTOR   11
#define NETLINK_NETFILTER   12  /* netfilter subsystem */
#define NETLINK_IP6_FW      13
#define NETLINK_DNRTMSG     14  /* DECnet routing messages */
#define NETLINK_KOBJECT_UEVENT  15  /* Kernel messages to userspace */
#define NETLINK_GENERIC     16
/* leave room for NETLINK_DM (DM Events) */
#define NETLINK_SCSITRANSPORT   18  /* SCSI Transports */
#define NETLINK_ECRYPTFS    19
#define NETLINK_RDMA        20
#define NETLINK_CRYPTO      21  /* Crypto layer */

can.h

/* particular protocols of the protocol family PF_CAN */
#define CAN_RAW     1 /* RAW sockets */
#define CAN_BCM     2 /* Broadcast Manager */
#define CAN_TP16    3 /* VAG Transport Protocol v1.6 */
#define CAN_TP20    4 /* VAG Transport Protocol v2.0 */
#define CAN_MCNET   5 /* Bosch MCNet */
#define CAN_ISOTP   6 /* ISO 15765-2 Transport Protocol */
#define CAN_NPROTO  7



```